irc/client_lookup_set.go

```diff
@@ -37,132 +36,108 @@ func ExpandUserHost(userhost string) (expanded string) {
-		cli := clients.ByNick[casefoldedName]
+		clients.RLock()
+		defer clients.RUnlock()
+		cli := clients.byNick[casefoldedName]
 		return cli
 	}
 	return nil
 }
 
-// Add adds a client to the lookup set.
-func (clients *ClientLookupSet) Add(client *Client, nick string) error {
-	nick, err := CasefoldName(nick)
-	if err != nil {
-		return err
-	}
-	clients.ByNickMutex.Lock()
-	defer clients.ByNickMutex.Unlock()
-	if clients.getNoMutex(nick) != nil {
-		return ErrNicknameInUse
+func (clients *ClientManager) removeInternal(client *Client) (removed bool) {
+	// requires holding ByNickMutex
+	oldcfnick := client.NickCasefolded()
+	currentEntry, present := clients.byNick[oldcfnick]
+	if present {
+		if currentEntry == client {
+			delete(clients.byNick, oldcfnick)
+			removed = true
+		} else {
+			// this shouldn't happen, but we can ignore it
+			client.server.logger.Warning("internal", fmt.Sprintf("clients for nick %s out of sync", oldcfnick))
+		}
 	}
-	clients.ByNick[nick] = client
-	return nil
+	return
 }
 
 // Remove removes a client from the lookup set.
-func (clients *ClientLookupSet) Remove(client *Client) error {
+func (clients *ClientManager) Remove(client *Client) error {
+	clients.Lock()
+	defer clients.Unlock()
+
 	if !client.HasNick() {
 		return ErrNickMissing
 	}
-	clients.ByNickMutex.Lock()
-	defer clients.ByNickMutex.Unlock()
-	if clients.getNoMutex(client.nick) != client {
-		return ErrNicknameMismatch
-	}
-	delete(clients.ByNick, client.nickCasefolded)
+	clients.removeInternal(client)
 	return nil
 }
 
-// Replace renames an existing client in the lookup set.
-func (clients *ClientLookupSet) Replace(oldNick, newNick string, client *Client) error {
-	// get casefolded nicknames
-	oldNick, err := CasefoldName(oldNick)
-	if err != nil {
-		return err
-	}
-	newNick, err = CasefoldName(newNick)
+// SetNick sets a client's nickname, validating it against nicknames in use
+func (clients *ClientManager) SetNick(client *Client, newNick string) error {
+	newcfnick, err := CasefoldName(newNick)
 	if err != nil {
 		return err
 	}
 
-	// remove and replace
-	clients.ByNickMutex.Lock()
-	defer clients.ByNickMutex.Unlock()
+	clients.Lock()
+	defer clients.Unlock()
 
-	oldClient := clients.ByNick[newNick]
-	if oldClient == nil || oldClient == client {
-		// whoo
-	} else {
+	clients.removeInternal(client)
+	currentNewEntry := clients.byNick[newcfnick]
+	// the client may just be changing case
+	if currentNewEntry != nil && currentNewEntry != client {
 		return ErrNicknameInUse
 	}
+	clients.byNick[newcfnick] = client
+	client.updateNickMask(newNick)
+	return nil
+}
 
-	if oldNick == newNick {
-		// if they're only changing case, don't need to remove+re-add them
-		return nil
+func (clients *ClientManager) AllClients() (result []*Client) {
+	clients.RLock()
+	defer clients.RUnlock()
+	result = make([]*Client, len(clients.byNick))
+	i := 0
+	for _, client := range(clients.byNick) {
+		result[i] = client
+		i++
```

Comment:
I'm guessing this making array then populating it this way is better performance-wise than using `append()` a million times, yeah? Just wanna confirm that's why we're doing it like this, since this way is a touch less intuitive initially.